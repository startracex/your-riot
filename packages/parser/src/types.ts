import type {
  IS_SELF_CLOSING,
  IS_SPREAD,
  JAVASCRIPT_TAG,
  STYLE_TAG,
} from "./constants.js";
type NodeTypes = {
  TAG: 1;
  ATTR: 2;
  TEXT: 3;
  CDATA: 4;
  COMMENT: 8;
  DOCUMENT: 9;
  DOCTYPE: 10;
  DOCUMENT_FRAGMENT: 11;
};
export type Expr = {
  unescape?: string;
  expressions?: {
    end?: number;
    text?: string;
    start?: number;
  }[];
  end?: number;
};

export type ParserState = Node &
  ExpressionContainer & {
    isVoid?: boolean;
    nodes?: ParserState[];
    pos?: number;
    data?: string;
    scryle?: ParserState | string;
    options?: {
      brackets?: [string, string];
      comments?: boolean;
    };
    close?: string;
    root?: ParserState;
    last?: ParserState;
    text?: string | ParserState;
    count?: number;
    name?: string;
    regexCache?: Record<string, RegExp>;
    [IS_SELF_CLOSING]?: boolean;
    builder?: Builder;
    stack?: ParserState[];
  };

type Store = {
  [STYLE_TAG]?: TagNode;
  [JAVASCRIPT_TAG]?: TagNode;
} & ParserState;

export type Builder = {
  compact: boolean;
  store: Store;
  get(): ParserOutput;
  push(node: ParserState): void;
  pushComment(store: Store, node: ParserState): void;
  closeTag(store: Store, node: ParserState): void;
  openTag(store: Store, node: ParserState): void;
  pushText(store: Store, node: ParserState): void;
  split(node: Store, source: string, start: number, pack: any): void;
  sanitise(node: Store, text: string, pack: any): void;
  attrs(attributes: Attribute[]): void;
};

export type Position = {
  start?: number;
  end?: number;
};

export type Expression = Position & {
  text?: string;
};

export type ExpressionContainer = {
  expressions?: Expression[];
  unescape?: string;
  parts?: string[];
};

export type Attribute = Position &
  ExpressionContainer & {
    name?: string;
    value?: string;
    valueStart?: number;
    [IS_SPREAD]?: boolean;
  };

export type Node = Position & {
  type?: number;
};

export type TagNode = Node & {
  type?: NodeTypes["TAG"];
  name: string;
  isRaw?: boolean;
  isCustom?: boolean;
  isVoid?: boolean;
  attributes?: Attribute[];
  selfclose?: boolean;
  nodes: Node[];
};

export type CommentNode = Node & {
  type: NodeTypes["COMMENT"];
};

export type TextNode = Node &
  ExpressionContainer & {
    type: NodeTypes["TEXT"];
    text: string;
  };

export type ParsedNode = TagNode | TextNode | CommentNode;

export type ParserOutput = {
  javascript: ScriptNode | null;
  css: StyleNode | null;
  template: TemplateNode | null;
};

export type TreeBuilder = {
  get(): ParserOutput;
  push(node: ParsedNode): void;
};

export type ParserOptions = {
  brackets?: [string, string];
  comments?: boolean;
  compact?: boolean;
};

export type ParserResult = {
  data: string;
  output: ParserOutput;
};

export type RiotNode = TemplateNode | TextNode | ScriptNode | StyleNode;

export type TemplateNode = TagNode & {
  nodes?: RiotNode[];
};

export type ScriptNode = TagNode & {
  name: "script";
  text?: TextNode;
};

export type StyleNode = TagNode & {
  name: "style";
  text?: TextNode;
};

export type Parser = (
  options: ParserOptions,
  treeBuilder?: TreeBuilder,
) => {
  parse: (data: string) => ParserResult;
};
