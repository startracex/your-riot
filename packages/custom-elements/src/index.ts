import {
  component,
  type RiotComponentWrapper,
  type RiotComponent,
  type RiotComponentFactoryFunction,
} from "@your-riot/riot";
import { defineProperty } from "@your-riot/utils/objects";

function moveChildren(source, target) {
  if (source.firstChild) {
    target.appendChild(source.firstChild);
    moveChildren(source, target);
  }
}

export function createElementClass(
  api: RiotComponentWrapper<RiotComponent>,
): typeof HTMLElement {
  const { css, exports, template } = api;
  const originalOnMounted = (exports as RiotComponent)?.onMounted;
  const tagImplementation = exports || {};

  return class extends HTMLElement {
    componentFactory?: RiotComponentFactoryFunction<RiotComponent>;
    component?: RiotComponent;

    constructor() {
      super();
      this.attachShadow({ mode: "open" });
      this.componentFactory = component({
        exports: {
          ...tagImplementation,
          onMounted: undefined,
        } as RiotComponent,
        template,
      });

      if (css) {
        const sheet = new CSSStyleSheet();
        sheet.replaceSync(css);
        this.shadowRoot?.adoptedStyleSheets.push();
      }
    }

    // on element appended callback
    connectedCallback() {
      this.component = this.componentFactory(this);

      // move the tag root html into the shadow DOM
      moveChildren(this.component.root, this.shadowRoot);

      // call the onmounted only when the DOM has been moved
      originalOnMounted?.apply(this.component, [
        this.component.props,
        this.component.state,
      ]);
    }

    // on attribute changed
    attributeChangedCallback(attributeName, oldValue, newValue) {
      if (!this.component) {
        return;
      }

      defineProperty(this.component, "props", {
        ...this.component.props,
        [attributeName]: newValue,
      });
      this.component.update();
    }

    // on element removed
    disconnectedCallback() {
      this.component.unmount();
    }

    // component properties to observe
    static get observedAttributes() {
      return (tagImplementation as any).observedAttributes || [];
    }
  };
}

export function define(
  name: string,
  api: RiotComponentWrapper<RiotComponent>,
  options?: ElementDefinitionOptions,
): void {
  if (!customElements.get(name)) {
    customElements.define(name, createElementClass(api), options);
  }
}

export default define;
