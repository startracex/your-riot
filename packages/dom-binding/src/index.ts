/**
 * Method used to bind expressions to a DOM node.
 * @example
 *
 * riotDOMBindings
 *  .template(
 *   `<div expr0> </div><div><p expr1> <section expr2></section></p>`,
 *   [
 *     {
 *       selector: '[expr0]',
 *       redundantAttribute: 'expr0',
 *       expressions: [
 *         {
 *           type: expressionTypes.TEXT,
 *           childNodeIndex: 0,
 *           evaluate(scope) {
 *             return scope.time;
 *           },
 *         },
 *       ],
 *     },
 *     {
 *       selector: '[expr1]',
 *       redundantAttribute: 'expr1',
 *       expressions: [
 *         {
 *           type: expressionTypes.TEXT,
 *           childNodeIndex: 0,
 *           evaluate(scope) {
 *             return scope.name;
 *           },
 *         },
 *         {
 *           type: 'attribute',
 *           name: 'style',
 *           evaluate(scope) {
 *             return scope.style;
 *           },
 *         },
 *       ],
 *     },
 *     {
 *       selector: '[expr2]',
 *       redundantAttribute: 'expr2',
 *       type: bindingTypes.IF,
 *       evaluate(scope) {
 *         return scope.isVisible;
 *       },
 *       template: riotDOMBindings.template('hello there'),
 *     },
 *   ]
 * )
 */
export { default as template } from './template.js'

/**
 * Bind a new expression object to a single DOM node.
 */
export { default as createBinding } from './binding.js'

/**
 * Create a single template expression.
 */
export { default as createExpression } from './expression.js'

/**
 * Object containing all the binding types.
 */
export { default as bindingTypes } from '@your-riot/utils/binding-types'

/**
 * Object containing all the expression types.
 */
export { default as expressionTypes } from '@your-riot/utils/expression-types'

export type * from './types.js'
export type { TemplateChunk } from './template.js'
