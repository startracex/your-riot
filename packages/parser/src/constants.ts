export const JAVASCRIPT_OUTPUT_NAME = 'javascript' as const
export const CSS_OUTPUT_NAME = 'css' as const
export const TEMPLATE_OUTPUT_NAME = 'template' as const

// Tag names
export const JAVASCRIPT_TAG = 'script' as const
export const STYLE_TAG = 'style' as const
export const TEXTAREA_TAG = 'textarea' as const

// Boolean attributes
export const IS_RAW = 'isRaw' as const
export const IS_SELF_CLOSING = 'isSelfClosing' as const
export const IS_VOID = 'isVoid' as const
export const IS_BOOLEAN = 'isBoolean' as const
export const IS_CUSTOM = 'isCustom' as const
export const IS_SPREAD = 'isSpread' as const
