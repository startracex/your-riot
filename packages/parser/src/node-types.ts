export const TAG = 1 as const /* TAG */
export const ATTR = 2 as const /* ATTR */
export const TEXT = 3 as const /* TEXT */
export const CDATA = 4 as const /* CDATA */
export const COMMENT = 8 as const /* COMMENT */
export const DOCUMENT = 9 as const /* DOCUMENT */
export const DOCTYPE = 10 as const /* DOCTYPE */
export const DOCUMENT_FRAGMENT = 11 as const /* DOCUMENT_FRAGMENT */
