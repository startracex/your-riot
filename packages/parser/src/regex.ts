/**
 * Matches the start of valid tags names; used with the first 2 chars after the `'<'`.
 * @const
 * @private
 */
export const TAG_2C: RegExp = /^(?:\/[a-zA-Z]|[a-zA-Z][^\s>/]?)/;
/**
 * Matches valid tags names AFTER the validation with `TAG_2C`.
 * $1: tag name including any `'/'`, $2: non self-closing brace (`>`) w/o attributes.
 * @private
 */
export const TAG_NAME: RegExp = /(\/?[^\s>/]+)\s*(>)?/g;
/**
 * Matches an attribute name-value pair (both can be empty).
 * $1: attribute name, $2: value including any quotes.
 * @private
 */
export const ATTR_START: RegExp = /(\S[^>/=\s]*)(?:\s*=\s*([^>/])?)?/g;

/**
 * Matches the spread operator
 * it will be used for the spread attributes
 */
export const SPREAD_OPERATOR: RegExp = /\.\.\./;
/**
 * Matches the closing tag of a `script` and `style` block.
 * Used by parseText fo find the end of the block.
 * @private
 */
export const RE_SCRYLE: {
  script: RegExp;
  style: RegExp;
  textarea: RegExp;
} = {
  script: /<\/script\s*>/gi,
  style: /<\/style\s*>/gi,
  textarea: /<\/textarea\s*>/gi,
};

// Do not touch text content inside this tags
export const RAW_TAGS: RegExp = /^\/?(?:pre|textarea)$/;
