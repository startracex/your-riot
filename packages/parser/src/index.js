import * as constants from './constants'
import * as nodeTypes from './node-types'
import parser from './parser'

export {
  /*
   * Factory function to create instances of the parser
   */
  parser,
  /**
   * The nodeTypes definition
   */
  nodeTypes,
  /**
   * Expose the internal constants
   */
  constants,
}

export default parser
