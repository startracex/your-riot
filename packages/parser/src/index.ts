import * as constants from './constants.js'
import * as nodeTypes from './node-types.js'
import parser from './parser.js'

export {
  /*
   * Factory function to create instances of the parser
   */
  parser,
  /**
   * The nodeTypes definition
   */
  nodeTypes,
  /**
   * Expose the internal constants
   */
  constants,
}

export default parser
export type * from './types.js'
