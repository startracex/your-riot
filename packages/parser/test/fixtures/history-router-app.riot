<history-router-app>
  <router base={props.base} onStarted={onStarted}>
    <route path="/">
      <p>Hello</p>
    </route>
    <route path="/goodbye/:user">
      <user name={route.params[0]}/>
    </route>
  </router>

  <script>
    import { Router, Route } from '../../src'
    import User from './user.riot'

    export default {
      components: {
        Router,
        Route,
        User
      },
      onStarted(currentRoute) {
        this.isRouterStarted = true
        this.currentRoute = currentRoute
      }
    }
  </script>
</history-router-app>