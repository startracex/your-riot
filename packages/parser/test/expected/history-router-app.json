{
  "template": {
    "type": 1,
    "name": "history-router-app",
    "start": 0,
    "end": 565,
    "isCustom": true,
    "nodes": [
      {
        "type": 1,
        "name": "router",
        "start": 23,
        "end": 220,
        "isCustom": true,
        "attributes": [
          {
            "name": "base",
            "value": "{props.base}",
            "start": 31,
            "end": 48,
            "expressions": [
              {
                "text": "props.base",
                "start": 36,
                "end": 48
              }
            ],
            "valueStart": 36,
            "parts": [
              "props.base"
            ]
          },
          {
            "name": "onStarted",
            "value": "{onStarted}",
            "start": 49,
            "end": 70,
            "expressions": [
              {
                "text": "onStarted",
                "start": 59,
                "end": 70
              }
            ],
            "valueStart": 59,
            "parts": [
              "onStarted"
            ]
          }
        ],
        "nodes": [
          {
            "type": 1,
            "name": "route",
            "start": 76,
            "end": 124,
            "isCustom": true,
            "attributes": [
              {
                "name": "path",
                "value": "/",
                "start": 83,
                "end": 91,
                "valueStart": 89,
                "parts": [
                  "/"
                ]
              }
            ],
            "nodes": [
              {
                "type": 1,
                "name": "p",
                "start": 99,
                "end": 111,
                "nodes": [
                  {
                    "type": 3,
                    "text": "Hello",
                    "start": 102,
                    "end": 107,
                    "parts": [
                      "Hello"
                    ]
                  }
                ]
              }
            ]
          },
          {
            "type": 1,
            "name": "route",
            "start": 129,
            "end": 208,
            "isCustom": true,
            "attributes": [
              {
                "name": "path",
                "value": "/goodbye/:user",
                "start": 136,
                "end": 157,
                "valueStart": 142,
                "parts": [
                  "/goodbye/:user"
                ]
              }
            ],
            "nodes": [
              {
                "type": 1,
                "name": "user",
                "start": 165,
                "end": 195,
                "isCustom": true,
                "attributes": [
                  {
                    "name": "name",
                    "value": "{route.params[0]}",
                    "start": 171,
                    "end": 193,
                    "expressions": [
                      {
                        "text": "route.params[0]",
                        "start": 176,
                        "end": 193
                      }
                    ],
                    "valueStart": 176,
                    "parts": [
                      "route.params[0]"
                    ]
                  }
                ],
                "isSelfClosing": true
              }
            ]
          }
        ]
      }
    ]
  },
  "css": null,
  "javascript": {
    "type": 1,
    "name": "script",
    "start": 224,
    "end": 543,
    "text": {
      "type": 3,
      "text": "\n    import { Router, Route } from '../../src'\n    import User from './user.riot'\n\n    export default {\n      components: {\n        Router,\n        Route,\n        User\n      },\n      onStarted(currentRoute) {\n        this.isRouterStarted = true\n        this.currentRoute = currentRoute\n      }\n    }\n  ",
      "start": 232,
      "end": 534
    }
  }
}